src/mscorlib/src/System/Text/EncodingForwarder.cs

```diff
@@ -0,0 +1,329 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.Contracts;
+
+namespace System.Text
+{
+    // Shared implementations for commonly overriden Encoding methods
+
+    internal static class EncodingForwarder
+    {
+        // We normally have to duplicate a lot of code between UTF8Encoding,
+        // UTF7Encoding, EncodingNLS, etc. because we want to override many
+        // of the methods in all of those classes to just forward to the unsafe
+        // version. (e.g. GetBytes(char[]))
+        // Ideally, everything would just derive from EncodingNLS, but that's
+        // not exposed in the public API, and C# prohibits a public class from
+        // inheriting from an internal one. So, we have to override each of the
+        // methods in question and repeat the argument validation/logic.
+
+        // These set of methods exist so instead of duplicating code, we can
+        // simply have those overriden methods call here to do the actual work.
+
+        // NOTE: This class should ONLY be called from Encodings that override
+        // the internal methods which accept an Encoder/DecoderNLS. The reason
+        // for this is that by default, those methods just call the same overload
+        // except without the encoder/decoder parameter. If an overriden method
+        // without that parameter calls this class, which calls the overload with
+        // the parameter, it will call the same method again, which will eventually
+        // lead to a StackOverflowException.
+
+        public unsafe static int GetByteCount(Encoding encoding, char[] chars, int index, int count)
+        {
+            // Validate parameters
+
+            Contract.Assert(encoding != null); // this parameter should only be affected internally, so just do a debug check here
```

Comment:
@jkotas IMHO that class is big enough already (2000+ lines of code), plus we'd have to name the methods something else (e.g. `DefaultGetByteCount`, `NLSGetByteCount`) since the overloads would conflict with the virtual overloads that were already there. I don't think taking it as a parameter and doing a null check is too big of a deal (that's what all extension methods have to do anyway, basically).